### PR TITLE
publisher confirm

### PR DESCRIPTION
**Pull request by cat-in-vacuum**
```go
func foo() {
	dialer, _ := amqpextra.NewDialer(
		amqpextra.WithURL("amqp://guest:guest@rabbitmq:5672/amqpextra"),
	)

	p, _ := dialer.Publisher(
		publisher.WithConfirmation(10),
	)

	if err = pub.Publish(msg); err != nil {
		log.Fatal(err)
	}

	dialer.Close()
        p.Close()
}
```

### CODE REVIEW

File: publisher/publisher.go [hunk 1]
```diff
@@ -50,6 +52,10 @@ type Publisher struct {
 	readyChs   []chan struct{}
 	unreadyChs []chan error
 
+	confirmation bool
+
+	conformationChs []AMQPChannel
```
Comment (makasim) [inline publisher/publisher.go:57]: define prop

```
confirmationCh chan <-chan error // resultCh
```

File: publisher/publisher.go [hunk 2]
```diff
@@ -151,6 +157,12 @@ func WithNotify(readyCh chan struct{}, unreadyCh chan error) Option {
 	}
 }
 
+func WihConfirmation(confirmation bool) Option {
```
Comment (makasim) [inline publisher/publisher.go:160]: typo `WihConfirmation` -> `WithConfirmation`

Comment (makasim) [reply publisher/publisher.go:160]:
```go
func WithConfirmation(buffer int) Option {
```

buffer is the conformationChs size
validate that buffer is greater than 1. 
make conformationChs with given size

File: publisher/publisher.go [hunk 3]
```diff
@@ -306,17 +318,33 @@ func (p *Publisher) channelState(conn AMQPConnection, connCloseCh <-chan struct{
 }
 
 func (p *Publisher) publishState(ch AMQPChannel, connCloseCh <-chan struct{}) error {
+
 	chCloseCh := ch.NotifyClose(make(chan *amqp.Error, 1))
 	chFlowCh := ch.NotifyFlow(make(chan bool, 1))
 
+	if p.confirmation {
```
Comment (makasim) [inline publisher/publisher.go:325]:
```
p.conformationChs != nil
```

File: publisher/publisher.go [hunk 4]
```diff
@@ -306,17 +318,33 @@ func (p *Publisher) channelState(conn AMQPConnection, connCloseCh <-chan struct{
...
+		err := ch.Confirm(false)
```
Comment (makasim) [inline publisher/publisher.go:326]: enable confirmation in `channelState`  just after successful init call

Comment (makasim) [reply publisher/publisher.go:326]: in case of err wait and retry

File: publisher/publisher.go [hunk 5]
```diff
@@ -379,6 +407,34 @@ func (p *Publisher) publish(ch AMQPChannel, msg Message) {
...
+		confirmationCh := ch.NotifyPublish(make(chan amqp.Confirmation, 1))
+
+		result := ch.Publish(
+			msg.Exchange,
+			msg.Key,
+			msg.Mandatory,
+			msg.Immediate,
+			msg.Publishing,
+		)
+
```
Comment (makasim) [inline publisher/publisher.go:426]:
```
if result != nil {
  return reply()
}

if p.confirmationCh != nil {
  select {
  case p.confirmationCh <- msg.resultCh:
  case <-p.ctx.Done():
    return reply(p.ctx.Err())
  }
}
```

**Comment by makasim** [comment]
add func for hanlding confimations: 

```
func (p) handleConfirmations() {
}
```

and run it in separate gorouting just like `go p.connectionState()`

**Review by makasim — CHANGES_REQUESTED** [review]

**Comment by makasim** [comment]
fixes https://github.com/makasim/amqpextra/issues/3

File: publisher/publisher.go [hunk 6]
```diff
@@ -305,18 +334,32 @@ func (p *Publisher) channelState(conn AMQPConnection, connCloseCh <-chan struct{
...
 	p.logger.Printf("[DEBUG] publisher ready")
 	p.notifyReady()
 	for {
 		select {
+		case c, ok := <-confirmation:
+			{
+				if !ok {
+					p.logger.Printf("[WARN] confirmation closed")
+				}
+				if c.Ack {
```
Comment (makasim) [inline publisher/publisher.go:351]:
```go
resultCh := <-confirmationCh

if c.Ack {
    resultCh <- nil
} else {
    resultCh <- fmt.Errorf("not delivered")
}
```

**Review by makasim — CHANGES_REQUESTED** [review]

**Review by makasim — CHANGES_REQUESTED** [review]

File: publisher/publisher.go [hunk 7]
```diff
@@ -305,7 +335,48 @@ func (p *Publisher) channelState(conn AMQPConnection, connCloseCh <-chan struct{
...
-func (p *Publisher) publishState(ch AMQPChannel, connCloseCh <-chan struct{}) error {
+func (p *Publisher) handlerConfirms(resultChCh chan chan error, confirmationCh chan amqp.Confirmation) {
+	for {
+		select {
+		case c, ok := <-confirmationCh:
+			if !ok {
+				p.logger.Printf("[WARN] confirmation closed")
+
+				p.emptyConfirmationCh(confirmationCh)
+				for {
```
Comment (cat-in-vacuum) [inline publisher/publisher.go:346]: Can resultCh <- amqp.ErrClosed be moved to separate method?

**Review by makasim — CHANGES_REQUESTED** [review]

**Comment by makasim** [comment]
e2e publish confirms enabled. publish ten messages and assert that result is nil

File: publisher/publisher_test.go [hunk 8]
```diff
@@ -2166,6 +2166,351 @@ func TestFlowControl(main *testing.T) {
...
+func TestPublisherConfirms(main *testing.T) {
+	main.Run("ErrorIfBufferLessThanOne", func(t *testing.T) {
+		defer goleak.VerifyNone(t)
+
+		connCh := make(chan *publisher.Connection)
+		_, err := publisher.New(connCh, publisher.WithConfirmation(1))
+		require.EqualError(t, err, "confirmation buffer size must be greater than 1")
+	})
+
+	main.Run("WaitIfChannelConfirmErrored", func(t *testing.T) {
```
Comment (makasim) [inline publisher/publisher_test.go:2178]: WaitRetryIfConfirmErrored

File: publisher/publisher_test.go [hunk 9]
```diff
@@ -2166,6 +2166,351 @@ func TestFlowControl(main *testing.T) {
...
+		assertReady(t, readyCh)
+
+		resultCh := make(chan error, 2)
+		p.Go(publisher.Message{ResultCh: resultCh})
+		p.Go(publisher.Message{ResultCh: resultCh})
+
+		confirmationCh <- amqp.Confirmation{Ack: true}
+		confirmationCh <- amqp.Confirmation{Ack: true}
+
+		close(confirmationCh)
```
Comment (makasim) [inline publisher/publisher_test.go:2264]: `defer close(confirmationCh)` just after `confirmationCh := make(chan amqp.Confirmation, 2)`

File: publisher/publisher_test.go [hunk 10]
```diff
@@ -2166,6 +2166,351 @@ func TestFlowControl(main *testing.T) {
...
+		p.Close()
+		assertClosed(t, p)
+
+		expected := `[DEBUG] publisher starting
+[DEBUG] publisher stopped
+`
+		require.Equal(t, expected, l.Logs())
+	})
+
+	main.Run("ResultNilIfConfirmed", func(t *testing.T) {
```
Comment (makasim) [inline publisher/publisher_test.go:2217]: AckConfirmTwoMessages

File: publisher/publisher_test.go [hunk 11]
```diff
@@ -2166,6 +2166,351 @@ func TestFlowControl(main *testing.T) {
...
+		expected := `[DEBUG] publisher starting
+[DEBUG] publisher ready
+[DEBUG] handle confirmation started
+[DEBUG] handle confirmation stopped
+[DEBUG] publisher stopped
+`
+		require.Equal(t, expected, l.Logs())
+	})
+
+	main.Run("ResultErroredIfNotConfirmed", func(t *testing.T) {
```
Comment (makasim) [inline publisher/publisher_test.go:2281]: NackConfirmTwoMessages

File: publisher/publisher_test.go [hunk 12]
```diff
@@ -2166,6 +2166,351 @@ func TestFlowControl(main *testing.T) {
...
+		p.Go(publisher.Message{ResultCh: resultCh})
+		p.Go(publisher.Message{ResultCh: resultCh})
+
+		confirmationCh <- amqp.Confirmation{Ack: false}
+		confirmationCh <- amqp.Confirmation{Ack: false}
+
+		close(connCloseCh)
+		close(confirmationCh)
+
+		require.EqualError(t, waitResult(resultCh, time.Millisecond*100), "not delivered")
```
Comment (makasim) [inline publisher/publisher_test.go:2332]: `not delivered` ->  `confirmation: nack`

File: publisher/publisher_test.go [hunk 13]
```diff
@@ -2166,6 +2166,351 @@ func TestFlowControl(main *testing.T) {
...
+
+		resultCh := make(chan error, 2)
+		p.Go(publisher.Message{ResultCh: resultCh})
+		p.Go(publisher.Message{ResultCh: resultCh})
+
+		confirmationCh <- amqp.Confirmation{Ack: false}
+		confirmationCh <- amqp.Confirmation{Ack: false}
+
+		close(connCloseCh)
+		close(confirmationCh)
```
Comment (makasim) [inline publisher/publisher_test.go:2330]: in this test you should not do it.

File: publisher/publisher_test.go [hunk 14]
```diff
@@ -2166,6 +2166,351 @@ func TestFlowControl(main *testing.T) {
...
+[DEBUG] publisher ready
+[DEBUG] handle confirmation started
+[DEBUG] handle confirmation stopped
+[DEBUG] publisher unready
+[DEBUG] publisher stopped
+`
+		require.Equal(t, expected, l.Logs())
+	})
+
+	main.Run("ResultErroredWhenConfirmationClosed", func(t *testing.T) {
```
Comment (makasim) [inline publisher/publisher_test.go:2348]: ConfirmationWhileConnectionClosed

File: publisher/publisher_test.go [hunk 15]
```diff
@@ -2166,6 +2166,351 @@ func TestFlowControl(main *testing.T) {
...
+[DEBUG] publisher ready
+[DEBUG] handle confirmation started
+[DEBUG] handle confirmation stopped
+[DEBUG] publisher unready
+[DEBUG] publisher stopped
+`
+		require.Equal(t, expected, l.Logs())
+	})
+
+	main.Run("AllMessagesGetConfirmationWhenConnectionClosed", func(t *testing.T) {
```
Comment (makasim) [inline publisher/publisher_test.go:2419]: ConfirmationWhileChannelClosed

**Review by makasim — CHANGES_REQUESTED** [review]

**Pull request merged**